src/main/java/chungus/Task.java

```diff
@@ -0,0 +1,405 @@
+     */
+    protected String format() {
+        return (isDone() ? "[X] " : "[ ] ") + desc;
+    }
+
+    /**
+     * Returns whether the task is complete.
+     * 
+     * @return Whether the task is complete.
+     */
+    public boolean isDone() {
+        return isDone;
+    }
+
+    /**
+     * Marks the task as complete.
+     */
+    public void setDone() {
+        isDone = true;
+    }
+
+    /**
+     * Marks the task as incomplete.
+     */
+    public void setNotDone() {
+        isDone = false;
+    }
+
+    /**
+     * Retrieves the raw description for the task.
+     * 
+     * @return The raw description.
+     */
+    public String desc() {
+        return desc;
+    }
+
+    @Override
+    public abstract String toString();
+
+    /**
+     * Serialies the task to make it suitable for storing to disk.
+     * 
+     * @return The serialized string.
+     */
+    public abstract String marshal();
+
+    /**
+     * Tries to deserialize a string to produce a task.
+     * 
+     * @param s The serialized string.
+     * @return The task.
+     * @throws TaskMarshalException If the string is invalid.
+     */
+    public static Task unmarshal(String s) {
+        assert s.length() > 1;
+        char typ = s.charAt(0);
+        switch (typ) {
+        case 'T':
+            return Todo.unmarshal(s);
+        case 'D':
+            return Deadline.unmarshal(s);
+        case 'E':
+            return Event.unmarshal(s);
+        default:
+            throw new TaskMarshalException(s);
+        }
+    }
+
+    /**
+     * Throws an exception if the condition is false. Just a convenience method.
+     * 
+     * @param cond The boolean condition.
+     * @param t    The exception to throw.
+     */
+    protected static void trueOrThrow(boolean cond, RuntimeException t) {
+        if (!cond) {
+            throw t;
+        }
+    }
+
+    public boolean hasTag(String tag) {
+        return tags.contains(tag);
+    }
+
+    public void addTags(String... tags) {
+        for (String tag : tags) {
+            this.tags.add(tag);
+        }
+    }
+
+    protected String chonkifyTags() {
+        StringBuilder sb = new StringBuilder();
+        for (String tag : tags) {
+            sb.append(Chonk.chonkify(tag));
+        }
+        return sb.toString();
+    }
+
+    protected static HashSet<String> dechonkifyTags(String s, int idx) {
+        Pair<String, Integer> dechonked;
+        HashSet<String> tags = new HashSet<>();
+
+        while (idx < s.length()) {
+            dechonked = Chonk.dechonkify(s, idx);
+            trueOrThrow(dechonked != null, new TaskMarshalException(s));
+            String tag = dechonked.first();
+            idx = dechonked.second();
+
+            tags.add(tag);
+        }
+
+        return tags;
+    }
+
+    public String[] getTags() {
+        return tags.stream().toArray(String[]::new);
+    }
+}
+
+/**
+ * A todo.
+ */
+class Todo extends Task {
```

Comment:
Perhaps put the different tasks in separate files? And then put them all in one separate package.